### ISSUE TITLE
pinyin

### ISSUE THREAD

**Issue by xcccdjsn** [Julow/Unexpected-Keyboard#585]
It would be perfect if it could support pinyin input, for me

**Comment by Julow** [comment]
This would requires word-correction/prediction ? The keyboard do not support that at the moment.

**Comment by Spike-from-NH** [comment]
Wikipedia says Pinyin is a way to express Chinese words using Roman letters. It should be doable with a custom layout and wouldn't require any new magic by Unexpected itself. There are "initials" (consonants) and "finals" (vowels), the latter modified by accents to select one of the four tones of Mandarin. It has been around for long enough that it obeyed the limitations of typewriters, so surely UNICODE has all the needed characters.

This might start with the QWERTY layout, with four swipes defined over each vowel. Or separate out the consonants from vowels, eliminate unused letters, and optimize based on native use.

**Comment by Spike-from-NH** [comment]
I'm looking at https://www.chinese-tools.com/tools/pinyin-editor.html —It provides a, e, i, o, u, and ü (umlaut), each with the macron, acute, grave, and caron accents, relying on UK or other soft keyboard for the initials.  **All these finals can already be produced by Unexpected.**

To set up: Swipe the gear logo, select Add keys to the keyboard, and check the **five** accents you will need. Back out to the keyboard. Now, for instance, swipe umlaut, swipe acute, and type u, which you will see now appears as: ǘ!

This is awfully cumbersome, so the final step is preparing a custom layout, where all 24 (6×4) finals are a single gesture, or even have their own key. @xcccdjsn, please reply if you need help with this.

**Comment by Spike-from-NH** [comment]
For example, modifying my QWERTY layout:
![20240311_165634](https://github.com/Julow/Unexpected-Keyboard/assets/126838489/abc50b47-0951-405f-a0fd-00aadc6ec86f)

**Comment by Spike-from-NH** [comment]
Here's the whole keyboard:
![20240311_214344](https://github.com/Julow/Unexpected-Keyboard/assets/126838489/8b27aa1f-eb2e-414b-9db9-ade0200cb2a8)
And here is the code (please remove ".txt"):
[Julow Pinyin.xml.txt](https://github.com/Julow/Unexpected-Keyboard/files/14566165/Julow.Pinyin.xml.txt)

**Comment by xcccdjsn** [comment]
@Spike-from-NH ,Thank you very much for answering my needs. Your solution is very good, but I am very sorry. I most likely expressed it wrong. Actually what I want to express is that you can input Chinese characters.

**Comment by xcccdjsn** [comment]
> This would requires word-correction/prediction ? The keyboard do not support that at the moment.



> This would requires word-correction/prediction ? The keyboard do not support that at the moment.

No, as long as you can type Chinese characters normally, it will be fine. Currently, my technology is not able to meet these requirements, and I can’t trouble you too much. I would be very grateful if it could support the output of Chinese characters. This would be a perfect product for me, and of course it is great now.

**Comment by Spike-from-NH** [comment]
> Actually what I want to express is that you can input Chinese characters.

@xcccdjsn, that is a lot to ask of a "lightweight keyboard." Perhaps a solution is to type Pinyin into Google Translate to get the Chinese characters. Also, the F-Droid repository has a phone app that lets you compose Chinese characters from the brush strokes.

**Comment by Spike-from-NH** [comment]
① That's a question for @Julow.
② Jules, perhaps copy my Pinyin XML above to your directory of useful experiments.
③ While working on this, I too "locked myself out of the house" by inserting a typo into an active layout. I used Samsung keyboard to fix the error, am unsure how I revived Unexpected. Is it worth coding a way to force a revert to QWERTY?—either by pop-up if the active layout is bad, or adding another button to the initial screen?

**Comment by JapanYoshi** [comment]
I looked up FOSS Chinese IMEs and found this. Is this good? https://rime.im/

**Comment by starlingnia** [comment]
what' the progress guys

I also deadly what a chinese support, unexpected keyboard have been perfect except for chinese support and suggestion function.

**Comment by HaleyHalcyon** [comment]
> what' the progress guys
> 
> I also deadly what a chinese support, unexpected keyboard have been perfect except for chinese support and suggestion function.

Seriously, can none of you people search for previous Issues?

Multiple contributors, including Julow himself, has stated that Chinese support is *not planned* for a lightweight keyboard app like Unexpected Keyboard.

Ask [fcitx5-android](https://github.com/fcitx5-android/fcitx5-android/issues/377) instead.

**Comment by starlingnia** [comment]
> > what' the progress guys进展如何？
> > I also deadly what a chinese support, unexpected keyboard have been perfect except for chinese support and suggestion function.我对中文客服非常失望，键盘本身很完美，除了中文客服和建议功能。
> 
> Seriously, can none of you people search for previous Issues?说真的，你们难道就没有人会去搜索一下以前的期刊吗？
> 
> Multiple contributors, including Julow himself, has stated that Chinese support is _not planned_ for a lightweight keyboard app like Unexpected Keyboard.包括 Julow 本人在内的多位贡献者都表示，像 Unexpected Keyboard 这样的轻量级键盘应用程序_没有计划_支持中文。
> 
> Ask [fcitx5-android](https://github.com/fcitx5-android/fcitx5-android/issues/377) instead.请咨询 [fcitx5-android](https://github.com/fcitx5-android/fcitx5-android/issues/377) 。


I should remind u that just in the post they referred to, there's a guy realize the same function for Korean. And u could found a lot of different languages support in ur keyboard settings, it's just that nobody achieve it for Chinese now. u should firstly read these comments in this issues then move to some recent issues, that's how it works

**Comment by Julow** [comment]
I'm working on spell checking (https://github.com/Julow/Unexpected-Keyboard/pull/1137), which requires handling large dictionaries. I think this is a very good starting point for implementing Chinese support. Especially for translating Pinyin, which requires a dictionary.
Anyone willing to implement this ?
